docs/storage/storage.md

```diff
@@ -0,0 +1,186 @@
+# Storage Design Notes
+## Status: (Very very)^N Draft
+### Author: Martin Smith
+
+## Tree Node Storage
+
+The node level of storage provides a fairly abstract tree model that is used to implement
+verifiable logs and maps. Most users will not need this level but it is important to know the
+concepts involved. The API for this is defined in `storage/tree_storage.go`. Related protos are in
+the `storage/storagepb` package.
+
+The model provides a versioned view of the tree. Revision numbers increase monotonically as the
+tree is mutated.
+
+### NodeIDs
+
+Nodes in storage are uniquely identified by a `NodeID`. This combines a tree path with
+a revision number. The path is effectively used as a bitwise subtree prefix in the tree. 
+
+The same `NodeID` objects are used by both logs and maps but they are interpreted differently.
+There are API functions that create them for each case. Mixing node ID types in API calls will
+give incorrect results.
+
+### Subtree Compaction
+
+As an optimization the tree is not stored as a set of raw nodes but at as a collection of subtrees.
+
+Currently, subtrees must be be a multiple of 8 levels deep (referred to as `strataDepth` in the
+code). Only the bottom level nodes the "leaves" of the subtree are physically stored.
+Intermediate subtree nodes are rehashed from the "leaves" when the subtree is loaded into memory.
+
+Subtrees are keyed by the `NodeID` of the root (effectively a path prefix) and contain the
+intermediate nodes for that subtree, identified by their suffix. These are actually stored in a
+proto map where the key is the suffix path. 
+
+Subtrees are versioned to support the access model for nodes described above. Node addresses within
+the model are distinct because the path to a subtree must be unique.
+
+When node updates are applied they will affect one or more subtrees and caching is used to increase
+efficiency. After all updates have been done in-memory the cache is flushed to storage so each
+affected subtree is only written once.
+
+Subtrees are helpful for reads because it is likely that many of the nodes traversed in
+Merkle paths for proofs are part of the same subtree. The number of subtrees involved in a path
+through a large tree from the leaves to the root is also bounded. For writes the subtree update
+batches what would be many smaller writes into one larger but manageable one.
+
+We gain space efficiency by not storing intermediate nodes. This is a big saving as it avoids
+storing entire tree levels, which get very large as the tree grows. This is magnified further as we
+store many versions of the tree.
+
```

Comment:
This saves effectively half the storage space, but we also lose out when (e.g. in maps) we're storing multiple subtree revisions (with the same prefix) just because one "leaf" differed...